### PR TITLE
new versions of WMCore dependency packages

### PR DESCRIPTION
**Pull request by vkuznet**
This PR is required for the following WMCore issues:
- https://github.com/dmwm/WMCore/issues/11205
- https://github.com/dmwm/WMCore/issues/11206
- https://github.com/dmwm/WMCore/pull/11248
- https://github.com/dmwm/WMCore/issues/11080

### CODE REVIEW

**Comment by vkuznet** [comment]
@muhammadimranfarooqi we need a new build for RPMs with changes presented in this PR. Please make them available.

**Comment by amaltaro** [comment]
@muhammadimranfarooqi Imran, please don't merge these yet.
@vkuznet Valentin, would you have 10min for a quick chat this afternoon? I just wanted to check with you on how you came up and tested these changes. Ping me on slack please.

**Comment by muhammadimranfarooqi** [comment]
-1 to 766bde334f2f1c6ed378a84de2687d263e0f603a on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_766bde3-20220822-195002.log" target="_blank">here</a>.

**Comment by vkuznet** [comment]
@amaltaro I don't have capacity to test on my own and rely on a build provided automatically. The build is triggered by PR, and original log had some errors. Once I get the log, I look at failures and fix them. This iteration will go on until we'll get new build. If successful I'll report it on WMCore ticket.

**Comment by muhammadimranfarooqi** [comment]
-1 to 2e71150de542211ba8c504ff98eabcc5abdf757b on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_2e71150-20220822-202003.log" target="_blank">here</a>.

**Comment by vkuznet** [comment]
test this please

**Comment by muhammadimranfarooqi** [comment]
+1 to 160602005f617bc9757f0eb233b401e7722285b3 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_1606020-20220822-203003.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to d59faba9d6bef420a8d485ecbf2643516d7f1967 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_d59faba-20220822-215502.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
@vkuznet Valentin, please check if the dependencies for the following packages are properly satisfied:
CherryPy, cx-Oracle, pyOpenSSL, pycurl, pylint, SQLAlchemy

I noticed that updates are missing for - at least - the pyOpenSSL package.

**Comment by vkuznet** [comment]
@amaltaro , I'll be checking dependencies one by one using your docker approach. The changes for pyOpenSSL are in and once I get confirmation from the build I'll continue with others.

**Comment by muhammadimranfarooqi** [comment]
+1 to 2cf5b3973c11847685f3411188ac119d829ec969 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_2cf5b39-20220823-141002.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to 0b491b3219e5206b226c42202a3d2dde0a3a5915 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_0b491b3-20220823-143002.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to a9d82ccbe060cbc67a162542b3cf5a9c34754218 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_a9d82cc-20220823-145503.log" target="_blank">here</a>.

**Comment by vkuznet** [comment]
@amaltaro , new build came clean. I checked all dependencies for CherryPy, cx-Oracle, pyOpenSSL, pycurl, pylint, SQLAlchemy, and they are in place. The only one failed to build was py3-cryptography. The new version fails with some weird message (see log https://cern.ch/cms-dmwm-builds/8032/test_0b491b3-20220823-143002.log). The error shows the following:
```
ERROR: Could not build wheels for cryptography which use PEP 517 and cannot be installed directly
```
I'm not sure what is actually means and I think it should be addressed separately. So far I reverted the upgrade for this specific package and our build is pass through.

Please let me know your feedback on this issue and how to proceed.

**Comment by amaltaro** [comment]
@vkuznet it might be that we need to update either pip or setuptools package as well, at least this is more or less what I found on google, e.g.:
https://github.com/pyca/cryptography/issues/5101

We need to look at the log and see what exactly fail. It looks like you jumped from version `3.2.1` to `37.0.4`, right? Have you looked into the release notes of the cryptography package to see if there is any breaking changes (quite possible with this major version upgrade).

UPDATE: which package requires cryptography==37.0.4?

**Comment by muhammadimranfarooqi** [comment]
-1 to 7f61569b83a675a09a48bfb65d87d949c7e99efc on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_7f61569-20220823-221002.log" target="_blank">here</a>.

**Comment by iarspider** [comment]
@amaltaro @vkuznet we can't update `py3-cryptography`: we use system OpenSSL (1.0.2k on SLC7), and newer versions of `py3-crypgraphy` require openssl 1.1.1-latest (they have dropped support for 1.0.2 in 35.0.0, and for 1.1.0 in 37.0.0).

**Comment by amaltaro** [comment]
@iarspider Ianna, thank you for this information. Looking into the cryptography release notes:
https://cryptography.io/en/latest/changelog/

it looks like even our current `3.2.1` should no longer support openssl 1.0.2 though.

Valentin, I don't know exactly how to proceed here. But maybe we need to try to build cryptography with the highest version we can (testing first this 35.x version).

**Comment by muhammadimranfarooqi** [comment]
-1 to e0e02a783a96fb56ef3e723850aa3ce8327c6a0a on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_e0e02a7-20220824-021003.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
-1 to 62af1508ece8605f8b3952e654fc0f6bdfef3277 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_62af150-20220824-150503.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to 427c6bbb47f99bc6f207085dd4698c2e43994334 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_427c6bb-20220824-153003.log" target="_blank">here</a>.

**Comment by vkuznet** [comment]
@amaltaro , I tried to use 3.4.8 version of cryprography and it fails therefore I reverted the change. I can try 3.3.X version too but I doubt it will make any sense. I suggest to keep current set of changes and address cryptography upgrade separately. Since you asked before, it is required for pyOpenSSL.

Please make a final call what to do with it and how to proceed. We have quite large chunk of changes in this PR and I'm in favor to merge it to make RPMs available and proceed with upgrade of CherryPy, pycurl, and others for WMCore needs.

**Comment by amaltaro** [comment]
@vkuznet can you please clarify in the initial description of this PR which WMCore GH issues this PR is supposed to fix?
I am planning to cut a new WMCore release today, so we can test with all these changes in as well. Thanks

**Comment by vkuznet** [comment]
done, I added relevant WMCore issues/PRs

File: py3-greenlet.spec
Comment (amaltaro) [inline py3-greenlet.spec]: @vkuznet can you please remove this line?

**Review by amaltaro — CHANGES_REQUESTED** [review]
Valentin, given that we have a bunch of changes already in place here, and that pyOpenSSL upgrade will apparently break things with its dependencies:
```
Collecting pyOpenSSL==22.0.0
  Downloading pyOpenSSL-22.0.0-py2.py3-none-any.whl (55 kB)
Collecting cryptography>=35.0
  Downloading cryptography-37.0.4-cp36-abi3-manylinux_2_24_x86_64.whl (4.1 MB)
Collecting cffi>=1.12
  Downloading cffi-1.15.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (442 kB)
Collecting pycparser
  Downloading pycparser-2.21-py2.py3-none-any.whl (118 kB)
Installing collected packages: pycparser, cffi, cryptography, pyOpenSSL
Successfully installed cffi-1.15.1 cryptography-37.0.4 pyOpenSSL-22.0.0 pycparser-2.21
```

I would suggest to remove these libraries above from this PR. Once we are confident that ALL these changes are working properly, we can start addressing the pyOpenSSL update.
Can you please also update your WMCore PR reflecting all these changes? Thanks

File: py3-greenlet.spec
Comment (vkuznet) [reply py3-greenlet.spec]: done

**Comment by vkuznet** [comment]
@amaltaro  I reverted back cffi, cryptography, pyOpenSSL and pycparser.

**Comment by muhammadimranfarooqi** [comment]
+1 to 05a7211df4eff09d2aa1a40b0e828eaecbb44697 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_05a7211-20220824-194003.log" target="_blank">here</a>.

File: py3-cffi.spec
Comment (amaltaro) [inline py3-cffi.spec]: Valentin, is there any reason for the downgrade of this package? If not, can you please revert it?

**Review by amaltaro — CHANGES_REQUESTED** [review]
@vkuznet thanks for reverting those. It looks like you missed cffi (I left a comment along the code).
Could you please squash all these commits to have a clean history (if rebasing, keep in mind you need to make it against `comp_gcc630`, not `master). I think the "merge" trigger will not automatically squash your commits, so we better make it manually.

File: py3-cffi.spec
Comment (vkuznet) [reply py3-cffi.spec]: sorry, it was my oversight, I changed minor version but forgot to revert last bit. Now it is done.

**Comment by vkuznet** [comment]
now, this PR is squashed too.

**Comment by muhammadimranfarooqi** [comment]
+1 to b1b80ed319035dfc7d5fcc51f5a24b697f3bbc08 on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_b1b80ed-20220824-213503.log" target="_blank">here</a>.

**Comment by muhammadimranfarooqi** [comment]
+1 to 36c072cb29a1968e4612d5c3a2b17d44ae30e23c on slc7_amd64_gcc630.<br>Build log can be found <a href="https://cern.ch/cms-dmwm-builds/8032/test_36c072c-20220824-215503.log" target="_blank">here</a>.

**Comment by amaltaro** [comment]
Thanks Valentin!

**Comment by amaltaro** [comment]
merge

**Comment by amaltaro** [comment]
Very short summary of the changes (and all their dependencies):
* cherrypy 17.4.0 -> 18.8.0
* sqlalchemy 1.3.3 -> 1.4.40
* cx-oracle 7.3.0 -> 8.3.0

**Comment by amaltaro** [comment]
FYI @belforte @mapellidario we should start tests of the WMCore services today/tomorrow and we will let you know if we find any problems and/or breaking changes with these new versions.

**Comment by belforte** [comment]
thanks @amaltaro . Is this a proper time to update DBSClient as well ? Or do we automatically get latest via pip anytime we buld a new CRAB rpm ? (sorry for may ignorance)

**Comment by amaltaro** [comment]
We have been using `4.0.8`:
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/py3-dbs3-client.spec

but I see there is one bug fix provided in tag `4.0.9`. We need to update the spec above in order to use the new version in our services (RPMs).

**Comment by belforte** [comment]
I guess it is my fault not having made a PR for that, since that bug fix was for a CRAB problem. Should I ?

**Comment by amaltaro** [comment]
Sure, go ahead Stefano! Keep in mind that it needs to be made against the branch `comp_gcc630`. Thanks!

**Comment by belforte** [comment]
branch `comp_gcc630` will never be forgotten !!

**Pull request merged**
